ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs

```diff
@@ -226,34 +230,70 @@ mkImmDB immDB decHeader decBlock encBlock epochInfo isEBB addHdrEnv err = ImmDB
   Getting and parsing blocks
 -------------------------------------------------------------------------------}
 
--- | Return the block corresponding to the given point, if it is part of the
--- ImmutableDB.
+-- | Return the block or header corresponding to the given point, if it is
+-- part of the ImmutableDB.
 --
 -- If we have a block at the slot of the point, but its hash differs, we
 -- return 'Nothing'.
 --
 -- If the point corresponds to some slot in the future, a
 -- 'ReadFutureSlotError' wrapped in a 'ImmDbFailure' is thrown.
-getBlockWithPoint :: forall m blk. (MonadCatch m, HasCallStack)
-                  => ImmDB m blk -> Point blk -> m (Maybe blk)
-getBlockWithPoint _  GenesisPoint = throwM NoGenesisBlock
-getBlockWithPoint db BlockPoint { withHash = hash, atSlot = slot } =
-    withDB db $ \imm -> do
-      mBlob <- fmap (uncurry (parse (decBlock db))) <$>
-        ImmDB.getBlockOrEBB imm slot hash
-      case mBlob of
-        Nothing         -> return $ Nothing
-        Just (Right b)  -> return $ Just b
-        Just (Left err) -> throwM $ err
-
-getBlockAtTip :: (MonadCatch m, HasCallStack)
-              => ImmDB m blk -> m (Maybe blk)
-getBlockAtTip db = do
+getBlockOrHeaderWithPoint
+  :: forall m blk b. (MonadCatch m, HasCallStack)
+  => ImmDB m blk
+  -> BlockOrHeader blk b
+  -> Point blk
+  -> m (Maybe b)
+getBlockOrHeaderWithPoint db blockOrHeader pt =
+    traverse deserialise =<<
+    getDeserialisableBlockOrHeaderWithPoint db blockOrHeader pt
+
+getDeserialisableBlockOrHeaderWithPoint
+  :: (MonadCatch m, HasCallStack)
+  => ImmDB m blk
+  -> BlockOrHeader blk b
+  -> Point blk
```

Comment:
Let's create a ticket to revisit this setup to see if we can unify some of these variants (it's conceivable that `BlockComponent` could help.)

(Can get rid of some simple wrappers already now perhaps.)

Reply:
I have created #1372 for this.